### ISSUE TITLE
machine learning algorithm

### ISSUE THREAD

**Issue by DonBeo** [scalanlp/breeze#567]
Hi,
is there an interest in the introduction of some machine learning algorithms?
If yes I can work on some of them.

Let me know

**Comment by dlwh** [comment]
I would be happy to consider them. You might look at the defunct Nak
library if you'd like. I'd definitely want to think through the client API
carefully, but I think it's reasonable.

**Comment by DonBeo** [comment]
I'll start to have a look at them. 

Regarding the API i think the best would be something like:
```

val lg = LogisticRegression(penalty=1)
val lgTrained = lg.fit(X, y) 
val hatY = lgTrained.predict(X)

```
and I would like  both `lg` and `lgTrained` immutable classes even I'm not sure that this is possible as they will definitively contain `DenseVectors` and similar

**Comment by haribageski** [comment]
I implemented back-propagation neural network based on your library. Please let me know if you are interested to take a look at it.

**Comment by DonBeo** [comment]
NN is in general a very wide field. I would say that apart from very simple algorithms as for example Extreme Learning Machines, NN should be in an independent library as for example DeepLearning4J. 

This is my opinion at least wait for dlnh's answer.

**Comment by dlwh** [comment]
I'm happy to take a look. I'm inclined to agree with @DonBeo that something
as complex as NN's should be a separate project. There's a lot of fiddling
and architecture stuff people do that I'd want to get right. (and the
available primitives in Breeze aren't super well suited to all the things
people want to do with NNs right now...)

A thing people frequently get wrong about ML is that it's really not about
specific techniques or procedures. It's about choosing a loss function, a
base set of features, (possibly jointly learned) transforms of those
features, and then a procedure for minimizing the loss (which most people
shouldn't be choosing). Any answer we come up with should address that.
TensorFlow seems to get a lot of this right.

-- David

**Comment by dlwh** [comment]
i'm going to close this. I'm currently trying to cut back on Breeze 1's scope
